### ISSUE TITLE
[lua] Can't throw objects: "error object is a table value"

### ISSUE THREAD

**Issue by RealyUniqueName** [HaxeFoundation/haxe#9161]
```haxe
class Main {
	static function main() {
		throw {};
	}
}
```
Runtime error (lua 5.3):
```
lua: (error object is a table value)
stack traceback:
        [C]: in function 'error'
        bin/test.lua:380: in field 'main'
        bin/test.lua:774: in main chunk
        [C]: in ?
```
Lua JIT complains about error value not being a string.
Doesn't matter if the object is an anonymous structure or a class instance.

**Comment by jdonaldson** [comment]
There doesn't seem to be a good way of handling this.  You can throw and catch arbitrary instances, but uncaught instances need to have a "message" that's a printable string (rather than the instance itself).  I'll see if there's some other workaround I could try.

**Comment by jdonaldson** [comment]
Ah, I see that I can add a __tostring function to a base metatable... and error will use that to generate a string.  It'll require some rework, but I can at least start in on it.

**Comment by RealyUniqueName** [comment]
Oh, wait. If it's just because of missing `__tostring` method, then you don't need to do anything. I'll handle it in #9124

**Comment by RealyUniqueName** [comment]
Alright, I thought it's just about adding a `__tostring` method in a Haxe class declaration, but looks like it should be something else :)
@jdonaldson, can you please point me where to add `__tostring`?

**Comment by jdonaldson** [comment]
Yeah, let me sketch some options:

1. We could just add a simple lua string printer to the base anonymous metatable on __tostring.  This would handle numbers/strings ok, but tables would show as "<table 0xfe384235>", etc.
2. We add the haxe Std.string printer to __tostring.  This means I would have to rework some of the initialization order, since Std.string relies on most of the other Haxe runtime primitives (Arrays, etc)

I'm more inclined to go with 2, since it seems to be more useful.

**Comment by RealyUniqueName** [comment]
I agree, 2 is much better.

**Comment by jdonaldson** [comment]
This was a bit of a rabbit hole, I had to move some of the initialization code for printing way up in the code generation stack.  I ended up rewriting a lot of the base helper methods to accommodate the change.  But, I think it's worth it to organize things a bit more rigorously.

All haxe-generated objects will have a `__tostring` metamethod that will help render instance information in the call stack.  I suppose there may be some people that want to customize this, but I'd rather see some use cases first before I try to support it.

There's still some base lua types that might cause problems when thrown... userdata types and the like.  But, those should be rare, at least within haxe generated lua code.
